### PR TITLE
Leverage Weight#count when size is set to 0

### PR DESCRIPTION
**Pull request by javanna**
We have removed shortcut total hit count with #89047 and later noticed a couple of benchmark regressions, which made us restore our shortcut total hit count mechanism. 

When not collecting hits (e.g. size=0) we can leverage Lucene skipping mechanism instead of our handmade shortcut total hit count, as Elasticsearch uses `TotalHitCountCollector` which calls `Weight#count`. The advantage of this is that it supports shortcutting for many more queries than the only 3 which our manual mechanism supports (match_all, term and field exists).

While at it, a few more tests are added to cover for situations that were not covered before.

### CODE REVIEW

**Comment by javanna** [comment]
run elasticsearch-ci/part-2

**Comment by dnhatn** [comment]
> With this change we reintroduce the shortcut total hit count method, and only use it when strictly necessary.

@javanna Would it be possible to revert #89047 first, then having the only adjustment in this PR? That would make the review easier. Thanks!

**Comment by javanna** [comment]
> > With this change we reintroduce the shortcut total hit count method, and only use it when strictly necessary.
> 
> @javanna Would it be possible to revert #89047 first, then having the only adjustment in this PR? That would make the review easier. Thanks!

That's what this change does, but without reinstating calling the method when size is set to 0 (EmptyTopDocsCollectorContext). What's the part that you find hard to follow?

**Comment by dnhatn** [comment]
@javanna The PR partially reverts #89047 and include additional changes. I had to compare this PR with #89047 to make sure we don't overlook anything.

**Comment by javanna** [comment]
@dnhatn this is ready ;)

File: server/src/main/java/org/elasticsearch/common/lucene/MinimumScoreCollector.java [hunk 1]
```diff
@@ -15,9 +15,15 @@
...
 import org.apache.lucene.search.ScoreMode;
 import org.apache.lucene.search.SimpleCollector;
+import org.apache.lucene.search.Weight;
 
 import java.io.IOException;
 
+/**
+ * Collector that wraps another collector and collects only documents that have a score that's greater or equal than the
+ * provided minimum score. Given that this collector filters documents out, it does and should not override {@link #setWeight(Weight)},
+ * as that may lead to exposing total hit count that does not reflect the filtering.
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/common/lucene/MinimumScoreCollector.java:25]: Thanks for adding this doc. I found it helpful :). Can we override and make the `setWeight` method final this class and have this comment in the method instead? It's totally optional.

**Review by dnhatn — APPROVED** [review]
LGTM. Thanks @javanna.

File: server/src/main/java/org/elasticsearch/common/lucene/MinimumScoreCollector.java (hunk 1 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/common/lucene/MinimumScoreCollector.java:25]: I will take these two out of this PR and open another one.
